### PR TITLE
fix(graphql): add root object version for dynamic field queries

### PR DESCRIPTION
**Pull request by unmaykr-aftermath**
## Description 

Adds the following attribute to `sui_graphql_rpc::types::object::Object`:
```rust
    /// Optional root parent object version if this is a dynamic field.
    ///
    /// This enables consistent dynamic field reads in the case of chained dynamic object fields,
    /// e.g., `Parent -> DOF1 -> DOF2`. In such cases, the object versions may end up like
    /// `Parent >= DOF1, DOF2` but `DOF1 < DOF2`. Thus, database queries for dynamic fields must
    /// bound the object versions by the version of the root object of the tree.
    ///
    /// Essentially, lamport timestamps of objects are updated for all top-level mutable objects
    /// provided as inputs to a transaction as well as any mutated dynamic child objects. However,
    /// any dynamic child objects that were loaded but not actually mutated don't end up having
    /// their versions updated.
    root_version: Option<u64>,
```

I tried making sure that this attribute is set correctly wherever an `Object` is created.

There's one scenario which still isn't covered: if the GQL query is rooted at a DOF, something like
```graphql
query DynamicFields($parent: SuiAddress!, $version: Int, $after: String) {
  object(address: $parent, version: $version) {
    dynamicFields(/* ... */) {
        // ...
    }
  }
}
```
where `$parent` is a DOF's object id, then the server has no way to know the root object's version, therefore it will only use `checkpoint_viewed_at` to bound the objects query, which of course varies with the current checkpoint number since it can't be set via the query. This can give incorrect results.

## Test plan 

How did you test the new or updated feature?

I need help with this. At Aftermath, we tested it on Testnet against a protocol and verified that queries rooted at the root parent object successfully returned two levels of DOFs even though the first level didn't change version because its contents didn't change in a transaction. However the state of that protocol has since evolved and the root object version we tested is outside the available range of the RPC.

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [x] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

**Comment by unmaykr-aftermath** [comment]
@wlmyng
@stefan-mysten
@emmazzz
@amnn

**Comment by wlmyng** [comment]
Hey @unmaykr-aftermath , greatly appreciate you opening this PR! I've just returned from PTO and will get to reviewing this later today

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (wlmyng) [inline crates/sui-graphql-rpc/src/types/object.rs:707]: nit: i think this can just be a standalone function so we don't need the prefix `Self::`

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (wlmyng) [inline crates/sui-graphql-rpc/src/types/object.rs:729]: nit: since this is what gets called elsewhere, I think we'll benefit from some doc comments explaining what we might return when calling `root_version()`.

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (wlmyng) [inline crates/sui-graphql-rpc/src/types/object.rs:1209]:
```suggestion
                // This conversion will use the object's own version as the `Object::root_version` if it's not a child object, but if it
                // is, it will be set to `None`, since we do not have any information on the `root_version` of the fetched dynamic object.
```

Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/object.rs:1209]: This makes me wonder if we should explicitly `infer_root_version` instead of deferring it to the function(s)?

**Review by wlmyng — COMMENTED** [review]
This looks good, thank you for tackling this bug. I just have two comments:
1. I believe we do not need `DynamicField.parent_version`, since it is set on `super_.super_.root_version()` per the ways we can instantiate a `DynamicField`
2. It would be great if we can introduce some comments on the functions that accept `root_version`, similar to what we have for `checkpoint_viewed_at`

As for testing, I should be able to branch off this and set up an e2e test

**Comment by unmaykr-aftermath** [comment]
> This looks good, thank you for tackling this bug. I just have two comments:
> 1. I believe we do not need `DynamicField.parent_version`, since it is set on `super_.super_.root_version()` per the ways we can instantiate a `DynamicField`
> 2. It would be great if we can introduce some comments on the functions that accept `root_version`, similar to what we have for `checkpoint_viewed_at`
> 
> As for testing, I should be able to branch off this and set up an e2e test

Thanks. I'll be able to tackle these next week

**Comment by unmaykr-aftermath** [comment]
@wlmyng I think I addressed all your comments in the last two commits. Let me know

**Comment by unmaykr-aftermath** [comment]
I wonder if we should error if the user roots a dynamic field query on a DOF. 

The issue in that scenario is that `Object::root_version` will return `None`. So imagine:
- the current checkpoint number is `100`
- the user queries an `object(...)`, which is a DOF of another object, at version `10`, which was created at checkpoint `80`
- the user asks for one or more dynamic fields (under `object(...) { ... }`)

The user would likely expect the dynamic fields to be the ones at the time version `10` of the DOF was created, i.e., at checkpoint `80`. However, since `Object::root_version` will be set to `None`, dynamic field queries will search up to the current checkpoint, `100`

**Comment by wlmyng** [comment]
> I wonder if we should error if the user roots a dynamic field query on a DOF.
> 
> The issue in that scenario is that `Object::root_version` will return `None`. So imagine:
> 
> * the current checkpoint number is `100`
> * the user queries an `object(...)`, which is a DOF of another object, at version `10`, which was created at checkpoint `80`
> * the user asks for one or more dynamic fields (under `object(...) { ... }`)
> 
> The user would likely expect the dynamic fields to be the ones at the time version `10` of the DOF was created, i.e., at checkpoint `80`. However, since `Object::root_version` will be set to `None`, dynamic field queries will search up to the current checkpoint, `100`

If I'm understanding your example correctly, then we actually want to root on the top-level object regardless of whether the object is a DOF/ child object or not? If so, I think we can address this by relaxing `Object::root_version`'s logic. If `root_version` is already stamped, we'll continue to use it. Otherwise, if it's not set, use the root object's version as the `root_version`, even if it is a child object.

Locally, I've set up an e2e test as follows: create Parent1, Child1, and Child2. Add Child1 as a child object to Parent1, and then add Child2 as a nested child object by borrowing Child1 from Parent1 and adding Child2 as a child to Child1. Finally, mutate nested Child 2.

The only change I made was to `infer_root_version`, to return `Some(native.version().into())`

With a graphql query like `object(Parent, version).dynamicFields.dynamicFields`, the results will be as follows:
Parent(5) -> Child1 (5) -> None // add child1 as child to parent1
Parent(6) -> Child1 (5) -> Child2 (6) // add child2 as a child to child1 by borrowing child1 from parent
Parent(7) -> Child1 (7) -> Child2 (6) // mutate child1
Parent(8) -> Child1 (7) -> Child2 (8) // mutate nested child2 by borrowing from child1

And then, if you were to query with Child1 as the root:
Child1 (5) -> None
Child1 (7) -> Child2 (6)

In comparison, with the existing `infer_root_version` implementation, the second request will always return Child2 (8)

**Comment by unmaykr-aftermath** [comment]
> And then, if you were to query with Child1 as the root:
> Child1 (5) -> None 
> Child1 (7) -> Child2 (6)
> 
> In comparison, with the existing `infer_root_version` implementation, the second request will always return Child2 (8)

Yeah, this seems to be as good as we can get without changing the implementation too much. There should be a caveat in some part of the documentation (not sure where) explaining that queries with child objects as root will view the state of that object version at the time of its **creation**. 

Otherwise, a user could do the following queries
1. Parent(6) -> Child1 (5) // add child2 as a child to child1 by borrowing child1 from parent
1. Child1 (5) -> None

expecting that they can break the query "Parent(6) -> Child1 (5) -> Child2 (6)" in parts. We actually ran into this scenario ourselves because we wanted to:
1. Get the contents of the Parent and a particular dynamic object (Child1)
2. Paginate over the dynamic fields of the dynamic object

However, that wasn't correct as we later found out. Unfortunately it's a very subtle, silent error that we could have overlooked and gone to production with

Ultimately, a better solution IMO would be do expose the `checkpoint_viewed_at` in the GQL parameters; quoting @amnn from Telegram:
> Eventually we want to expose this via an extra parameter to these queries, like so:
>
> ```graphql
> query ($objectId: SuiAddress!, $checkpoint: Int) {
>   object(address: $objectId, latestAt: $checkpoint) {
>     # ...
>   }
> }
> ```
>
> ```graphql
> query ($tableId: SuiAddress!, $checkpoint: Int) {
>   owner(address: $tableId, latestAt: $checkpoint) {
>     dynamicFields { 
>       # ...
>     }
>   }
> }
> ```

**Comment by wlmyng** [comment]
> There should be a caveat in some part of the documentation (not sure where) explaining that queries with child objects as root will view the state of that object version at the time of its creation.

I think you're right in suggesting that we ought to return an error if someone is using the object field rooted on a DOF. The error can point the caller to use the owner field instead.

Meanwhile, on the Query.owner field, we can introduce an additional optional root_version parameter, so that someone can provide the nested DOF that owns another DOF, and the proper root_version to fix to a point in time.

**Comment by unmaykr-aftermath** [comment]
> I think you're right in suggesting that we ought to return an error if someone is using the object field rooted on a DOF, due to the inconsistency between a caller expecting that they can break the query "Parent(6) -> Child1 (5) -> Child2 (6)" in parts and what they are actually likely to get back
> 
> The error can point the caller to use the owner field instead.
> 
> Meanwhile, on the Query.owner field, we can introduce an additional optional root_version parameter, so that someone can provide the address of some nested DOF that owns another DOF, and the proper root_version to fix to a point in time. I think that will also more accurately bound the history than setting a checkpoint

I like the ideas. The question is what is in scope for this PR. Should I add the error here and leave the Query.owner extension for later?

**Comment by wlmyng** [comment]
> I like the ideas. The question is what is in scope for this PR. Should I add the error here and leave the Query.owner extension for later?

I suppose until we allow someone to provide a `root_version`, we should disallow someone from being able to root dynamic fields on a DOF on either the `object` or `owner` field; as in, it would only be possible to root on a DOF if a root_version is also provided, to disambiguate the possibility of returning historical dat at dof's mutation vs from the perspective of breaking up a nested query.

Would you be interested in tackling the Query.owner extension? I think it can be done in a follow-up PR

**Comment by unmaykr-aftermath** [comment]
> I suppose until we allow someone to provide a `root_version`, we should disallow someone from being able to root dynamic fields on a DOF on either the `object` or `owner` field; as in, it would only be possible to root on a DOF if a root_version is also provided, to disambiguate the possibility of returning historical dat at dof's mutation vs from the perspective of breaking up a nested query.
> 
> Would you be interested in tackling the Query.owner extension? I think it can be done in a follow-up PR

I'm interested, yes. It would solve some problems for us as well.

So I'll add some sort of error when someone requests dynamic fields from a child object without a set `root_version`

**Comment by unmaykr-aftermath** [comment]
Hey @wlmyng, I noticed some weird scenarios while implementing our original idea so I came up with another solution in the latest commit. Here's the reasoning.

I went back and thought about the requirements:
1. the root object version should be propagated when we ask for a specific object version and its DF tree
1. asking for a specific child object version and its DF tree should error since it's ambiguous (the point about breaking up the query "Parent(6) -> Child1 (5) -> Child2 (6)")
1. asking for the DF tree of any `Owner` at the latest checkpoint should be allowed by just omitting any object version requirement at the root query

Hence I figured the best solution would be to propagate an `at_version` field attached to objects:
```rust
    /// Whether a specific version of this object was queried.
    ///
    /// If so, dynamic field queries from this object will fail if `root_version` isn't set.
    /// If not, dynamic field queries will use `checkpoint_viewed_at` as a bound **if**
    /// `root_version` isn't set.
    at_version: bool,
```

If the user didn't signal they wanted a part of the Sui state under a specific object version, we allow `root_version` of that `Object` to be none and hence everything bellow it is bounded by the latest checkpoint version. This allows queries like this to succeed:
```mermaid
flowchart LR
Query(("`Query`")) -- owner --> Owner(("`Owner`"))  -- dynamic_object --> MoveObject(("`MoveObject`")) -- dynamic_field --> DynamicField(("`DynamicField`"))
```

However if they did set a specific object version, we stop them if a dynamic field tree in the query is rooted at a child object's version. This can happen directly at `Query.object` but also indirectly if the user requests dynamic fields from an object in `ObjectChange.inputState`, for example.

Looking at the schema in [GraphQL Voyager](https://graphql-kit.com/graphql-voyager/), it seems the only remaining problematic path is if the user requests `Query.objects` with `ObjectFilter::object_keys` set and at least one of those being a child object, since the current solution only sets the `at_version` flag in paths that use `Loader<HistoricalKey>`. I can solve that in the next PR I think, using a `HashSet` of the object ids in `ObjectFilter::object_keys` and setting the `at_version` flag on objects found.

Of course, adding `owner(root_version: Int)` in a follow-up PR is still going to be super useful too.

**Comment by wlmyng** [comment]
Hey @unmaykr-aftermath greatly appreciate you looking into these scenarios!
Sorry for getting back to you so late, I had to sit on this for a bit. Perhaps the best approach is to not special-case a child object being used as the root on the object path, and instead just proceed as normal. 
Otherwise, it looks like we'll produce inconsistent results depending on how a caller hits `childObject.dynamicFields` - from running your latest changes locally, an explicit top-level query returns an error, but if the scenario occurs indirectly, say through `objects(filter)` or `ObjectChange.inputState`, we will still render the results, anchored at the historical state at the child object's version bump. On the other hand, if the intention is to raise an error always, I suspect that this will cause significant friction, since the entire query will error out. Someone may have unintentionally and unwillingly queried for child objects in their `objects.dynamicFields` query, and it would be difficult to modify their query to avoid this scenario.
This should also simplify the implementation, since we don't need to edge case all instances of child object -> df. Consequently, we can phase out the new `at_version` field and relax `infer_root_version`.
Then, separately, the `Query.owner(address, rootVersion)` API will explicitly address your use case of needing to view a child object at a particular moment in time.

I know I had agreed earlier that we should return an error, so apologies for going back on that and any consequent churn. Returning the same data at any level of a graphql request allows us to deliver a consistent response and minimizes any potential pain from someone unaware of child-object-rooted dynamic fields query intricacies. Meanwhile, power users can leverage the new parameter on `owner` for their historical queries.

**Comment by unmaykr-aftermath** [comment]
> Perhaps the best approach is to not special-case a child object being used as the root on the object path, and instead just proceed as normal. 
> 
> This should also simplify the implementation, since we don't need to edge case all instances of child object -> df. Consequently, we can phase out the new `at_version` field and relax `infer_root_version`. Then, separately, the `Query.owner(address, rootVersion)` API will explicitly address your use case of needing to view a child object at a particular moment in time.

So the ideia is to propagate the root version if we have it and if not (if starting from the child object directly for instance)
to set the root version as the current object's root version, correct?

I'll work on that and then on `Query.owner(address, rootVersion)` in a subsequent PR.

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (wlmyng) [inline crates/sui-graphql-rpc/src/types/object.rs:889]:
```suggestion
    /// `root_version` represents the version of the root object in some nested chain of dynamic fields. This should typically be left `None`, unless the object(s) being resolved is a dynamic field, or if `root_version` has been explicitly set for this object. 
    /// If None, then we use [`version_for_dynamic_fields`] to infer a root version to then
```


On fetching object, `try_from_stored_history_object` will set the object's `object_version` as `root_version`
- this is passed as `parent_version` to `DynamicField::query` or `::paginate`
- if it's a `df`, it will already be bounded by the `parent_version`
- if it's a `dof`, then when we resolve its `value`, we pass the `root_version` inherited to the fetched `MoveObject`
- when resolving the `MoveObject`'s `DynamicField`, this cycle is perpetuated
However, if we break this chain, then the `root_version` resets until we try to fetch a `DynamicField` again

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (wlmyng) [inline crates/sui-graphql-rpc/src/types/object.rs:947]: This might make more sense as a doc comment for `fn value` of `DynamicField`, something like "If the `value` is a child object, its contents come from the latest version that is at most equal to its parent object's version."

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (wlmyng) [inline crates/sui-graphql-rpc/src/types/object.rs:956]: Is there any reason why we can't use `Some(native.version())` instead? I suppose because `Immutable` freezes the object version so it's no longer useful as a means to give us greater historical granularity (compared to just using `checkpoint_sequence_number`)

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/object.rs:947]: If that makes sense, you'll need to run `snapshot_tests.rs` to regenerate the schema and update the snapshot test

**Review by wlmyng — APPROVED** [review]
Please take a look at the notes regarding doc comments, but otherwise this looks good to me! Thanks again for tackling this, seems like a very tricky matter to correctly thread the `root_version` through objects

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (unmaykr-aftermath) [reply crates/sui-graphql-rpc/src/types/object.rs:956]: Yeah, that can be simpler. It should work assuming:
* the 'freeze' operation bumps the version of the parent object; it should since we're acquiring it by value? (write lock)
* removing a DOF from its parent and freezing it (if that is possible) always bumps its version

That last point is important since we may have the "Parent(6) -> Child1 (5) -> Child2 (6)" scenario. What happens to Child1's version if we remove it from Parent and freeze it without changing its contents?

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (unmaykr-aftermath) [reply crates/sui-graphql-rpc/src/types/object.rs:889]: Yeah that sounds correct

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (unmaykr-aftermath) [reply crates/sui-graphql-rpc/src/types/object.rs:947]: This is what I came up for `DynamicField.value`:
```
     /// The actual data stored in the dynamic field.
     /// The returned dynamic field is an object if its return type is MoveObject,
-    /// in which case it is also accessible off-chain via its address.
+    /// in which case it is also accessible off-chain via its address and the latest
+    /// version that is at most equal to its parent object's version.
```
Good to go?

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (unmaykr-aftermath) [reply crates/sui-graphql-rpc/src/types/object.rs:956]: Otherwise we have this scenario where querying Child1 (5) will not return Child1 (6) and the former has owner `Immutable` so we don't have a direct path to its previous root's version (at the time Child1 was frozen) that we can use in the upcoming `Query.owner(address, rootVersion)`

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/object.rs:947]: Apologies in advance for bikeshedding 😅 but it's possible for someone to access the child object directly off-chain via its address and at any version, so this is not entirely accurate. Proposing...

The returned dynamic field is an object if its return type is `MoveObject`, and its contents will be from the latest version that is at most equal to its parent object's version. Otherwise, it is also accessible off-chain via its address.

File: crates/sui-graphql-rpc/schema/current_progress_schema.graphql
Comment (amnn) [inline crates/sui-graphql-rpc/schema/current_progress_schema.graphql:895]: This currently reads as if the value is accessible via its address if its return type is *not* `MoveObject`, but it's the other way around, suggesting:

```suggestion
	The returned dynamic field is an object if its return type is `MoveObject`,
	in which case it is also accessible off-chain via its address. Its contents
	will be from the latest version that is at most equal to its parent object's
	version.
```

**Comment by unmaykr-aftermath** [comment]
@amnn Implemented your suggestions

File: crates/sui-graphql-rpc/src/types/dynamic_field.rs
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/dynamic_field.rs:97]: I think this comment still needs updating.

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/object.rs:1226]: nit: this comment needs to updated as well now, right? Because we'll always set to the object's own version if this is `None`.

**Review by amnn — COMMENTED** [review]
Excellent, thank you @unmaykr-aftermath ! I think we're now waiting on the @wlmyng to land the tests on top of your commits and this will be good to go -- I will leave it in both your capable hands, it looks good to me.

File: crates/sui-graphql-rpc/src/types/object.rs
Comment (unmaykr-aftermath) [reply crates/sui-graphql-rpc/src/types/object.rs:1226]: Good catch

File: crates/sui-graphql-rpc/src/types/dynamic_field.rs
Comment (unmaykr-aftermath) [reply crates/sui-graphql-rpc/src/types/dynamic_field.rs:97]: Yeah I think I accidentally undid it because I committed your suggestion directly on the schema file instead of on the source code and then running tests

**Comment by wlmyng** [comment]
@unmaykr-aftermath I'm going to land some tests covering the scenarios we discussed in this PR on top of your commits. Once they pass, we should be good to merge this. Thanks again!

**Comment by wlmyng** [comment]
@unmaykr-aftermath I'm not that good with github foo, this is what I came up with to land the commits in this branch 😅 https://github.com/AftermathFinance/sui-public/pull/1

**Comment by unmaykr-aftermath** [comment]
@wlmyng Merged your PR. There are conflicts here now that I don't know how to resolve. Can you solve them using the GitHub UI?

**Comment by amnn** [comment]
@wlmyng, you can push directly onto the branch for this PR to add commits (add the AftermathFinance version of the repo as a remote, and then push to the branch there). Github allows this because the branch is associated with a PR for a repo owned by an org that you are a member of.

**Comment by wlmyng** [comment]
Hey @unmaykr-aftermath , truly sorry for all the back and forth. I believe what @amnn proposed I should do is untenable due to this issue https://github.com/orgs/community/discussions/5634 (come on github!!!) so it looks like I'll have to create another PR pointed against this branch. Apologies for the churn! I can't use the web editor directly to resolve this conflict since it'll leave behind some unwanted artifacts. Instead I'll need to rebase the branch on main and resolve any conflicts there. It'll be up in a bit!

**Comment by unmaykr-aftermath** [comment]
> Hey @unmaykr-aftermath , truly sorry for all the back and forth. I believe what @amnn proposed I should do is untenable due to this issue https://github.com/orgs/community/discussions/5634 (come on github!!!) so it looks like I'll have to create another PR pointed against this branch. Apologies for the churn! I can't use the web editor directly to resolve this conflict since it'll leave behind some unwanted artifacts. Instead I'll need to rebase the branch on main and resolve any conflicts there. It'll be up in a bit!

All good. Just lmk

**Comment by wlmyng** [comment]
@unmaykr-aftermath while I have something in this PR https://github.com/AftermathFinance/sui-public/pull/2 I think it makes a mess of the commit history, and I'm not quite sure what this PR will look like when we merge that one ...  Wondering if instead you'd be able to sync the fork to `MystenLabs:main` and rebase, or if that is not desirable, to rebase this branch onto latest main from `MystenLabs`?

You'll likely run into some merge conflicts - feel free to accept incoming changes or resolve them however you'd like. It'll also complain about `query_latest_dynamic_field` since we need to update the call to `let gql_object =
            Object::try_from_stored_history_object(history_object, checkpoint_viewed_at, None)?;`, but I can pick this up along with any changes to `.exp` files needed from rerunning `sui-graphql-e2e-tests`. 

But if you do have things set up locally, you can also rerun the e2e tests on your end with `env UB=1 cargo nextest run --package sui-graphql-e2e-tests --features pg_integration`

**Comment by unmaykr-aftermath** [comment]
> But if you do have things set up locally, you can also rerun the e2e tests on your end with `env UB=1 cargo nextest run --package sui-graphql-e2e-tests --features pg_integration`

Hey @wlmyng I wasn't able to run tests properly since I ran into errors like this:
```
---- run_test::epoch/system_state.move ----
test panicked: Timeout waiting for graphql server to start: Elapsed(())


failures:
    run_test::epoch/system_state.move

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 62 filtered out; finished in 15.81s
```

I decided then to merge `main` into this branch and manually cherry-pick your changes. I checked that the e2e tests and graphql bin match using
```
git remote add -f wlmyng git@github.com:wlmyng/sui-public.git
git remote update
git diff wlmyng/resolve-conflicts-unmaykr-fix-dof crates/sui-graphql-e2e-tests/ crates/sui-graphql-rpc
```

Can you check that all tests pass?

**Comment by wlmyng** [comment]
@unmaykr-aftermath - looks like we are good to go. 🚢  it!

**Comment by unmaykr-aftermath** [comment]
Who can look at the failing checks and eventually merge it? :eyes:

**Comment by wlmyng** [comment]
I think we can ignore the Vercel checks, they are not required. You should have the ability to squash and merge. But if not, I'm happy to press the button for you

**Comment by unmaykr-aftermath** [comment]
> I think we can ignore the Vercel checks, they are not required. You should have the ability to squash and merge. But if not, I'm happy to press the button for you

I don't have that ability unfortunately. You can do the honors!

**Pull request merged**
